### ISSUE TITLE
Entities are not created in the correct order and caused error "Cannot add or update a child row: a foreign key constraint fails"

### ISSUE THREAD

**Issue by zorji** [typeorm/typeorm#2853]
I received error "Cannot add or update a child row: a foreign key constraint fails" when I have a non-nullable primary column in the relation.

**Issue type:**

[ ] question
[x] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[x] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[ ] `latest`
[x] `@next`
[x] `0.2.7` It happens to both @next and 0.2.7

**Steps to reproduce or a small repository showing the problem:**

I have 3 entities: `House`, `Room` and `Bed`.
`House` and `Bed` have a single column ID. `Room` has a composite ID of `roomNumber` and `House.id`. `Room` also has a nullable reference to `Bed`. Both `House` and `Bed` have no reference to `Room`.

Here is the minimal code to explain the issue
```ts
@Entity()
export class House {
  @PrimaryColumn('char', {length: 16})
  id: string
  @OneToMany(type => Room, room => room.house, {cascade: true, eager: true})
  rooms: Room[]
}

@Entity()
export class Bed {
  @PrimaryColumn('char', {length: 16})
  id: string
}

@Entity()
export class Room {
  @PrimaryColumn()
  roomNumber: number
  @ManyToOne(type => House, house => house.rooms, {
    primary: true,
  })
  house: House

  @OneToOne(type => Bed, {cascade: true, eager: true})
  @JoinColumn()
  bed: Bed
}

const connection = createConnection({...})
const record = new House({
  id: 'house_1',
  rooms: [
    new Room({
    roomNumber: 1,
      bed: new Bed({
        id: 'bed_1',
      }),
    }),
  ],
})
await connection.getRepository(House).save(record)
```

In my test, when the `@ManyToOne` relation in `Room` contains `primary: true` or `nullable: false`, the script would fail because it always tried to inserting the new `Room` before `Bed` was created. If I remove `primary: true`, it knows it should insert `Bed` before a `Room` is created and work as expected.

By debugging I found that it's `SubjectTopoligicalSorter.sort("insert")` sorting the subjects incorrectly (source code around `typeorm/src/persistence/SubjectExecutor.ts:115`)

Below is the full code to reproduce the issue. 
```ts
import 'reflect-metadata'
import {
  Column,
  ConnectionOptions,
  createConnection,
  Entity,
  JoinColumn,
  ManyToOne,
  OneToMany,
  OneToOne,
  PrimaryColumn,
} from 'typeorm'

@Entity()
export class House {
  @PrimaryColumn('char', {length: 16})
  id: string
  @OneToMany(type => Room, room => room.house, {cascade: true, eager: true})
  rooms: Room[]

  constructor(data: {
    id: string
    rooms: Room[]
  }) {
    Object.assign(this, data)
  }
}

@Entity()
export class Bed {
  @PrimaryColumn('char', {length: 16})
  id: string

  constructor(data: {
    id: string
  }) {
    Object.assign(this, data)
  }
}

@Entity()
export class Room {
  @PrimaryColumn()
  roomNumber: number
  @ManyToOne(type => House, house => house.rooms, {
    primary: true,
  })
  house: House

  @Column()
  north: boolean = true

  @OneToOne(type => Bed, {cascade: true, eager: true})
  @JoinColumn()
  bed: Bed

  constructor(data: {
    roomNumber: number
    bed?: Bed
  }) {
    Object.assign(this, data)
  }
}

(async () => {
  const config: ConnectionOptions = {
    type: 'mysql',
    host: 'localhost',
    port: 3306,
    username: 'root',
    password: 'secret',
    synchronize: false,
  }
  const connection0 = await createConnection(config)
  await connection0.createQueryRunner().query(`DROP DATABASE IF EXISTS mydb;`)
  await connection0.createQueryRunner().query(`CREATE DATABASE mydb CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;`)
  await connection0.close()

  const connection = await createConnection({
    ...config,
    database: 'mydb',
    entities: [
      House, Room, Bed,
    ],
    synchronize: true,
    logging: true,
  })
  try {
    const record = new House({
      id: 'house_1',
      rooms: [
        new Room({
          roomNumber: 1,
          bed: new Bed({
            id: 'bed_1',
          }),
        }),
      ],
    })
    await connection.getRepository(House).save(record, {
      transaction: false
    })
  } catch (error) {
    console.log('<b>Error: ', JSON.stringify(error, null, 2), '</b>')
    console.log('Error: ', error)
  }
  await connection.close()
})().catch(async error => {
  console.log('<b>Error: ', JSON.stringify(error, null, 2), '</b>')
  console.log('Error: ', error)
})
```

**Comment by vixeven** [comment]
I had a problem with saving new items: `Cannot add or update a child row: a foreign key constraint fails` and that was because I had both on my entity: `@ManyToOne` and `@Column` definitions. Apparently, you cannot have both. 🤷‍♂️

**Comment by ZBAGI** [comment]
`subjects` metadata are fetched reclusively, starting from farthest child ending with root parent. Then [SubjectExecutor](https://github.com/typeorm/typeorm/blob/8a5e6715e2d32da22c2fa71a14a7cf1fe897a159/src/persistence/SubjectExecutor.ts#L248) builds the query based on this order. Simply reversing this order fixes all the bulk insertion issues in mysql. Here is bugfix you can use without modifying any `typeorm` package files:
```ts
import { Subject } from "typeorm/persistence/Subject";
import { SubjectExecutor } from "typeorm/persistence/SubjectExecutor";

// Bugfix for bulk insertion order.
// Read more: https://github.com/typeorm/typeorm/issues/2852
(<any>SubjectExecutor.prototype).groupBulkSubjects = function (subjects: Subject[], type: "insert" | "delete"): [{ [key: string]: Subject[] }, string[]] {
	const group: { [key: string]: Subject[] } = {};
	const keys: string[] = [];
	const groupingAllowed = type === "delete" || (<any>this).queryRunner.connection.driver.isReturningSqlSupported();

	// Reversing order - ie. actual bugfix bellow
	subjects.reverse().forEach((subject, index) => {
		const key = groupingAllowed || subject.metadata.isJunction ? subject.metadata.name : subject.metadata.name + "_" + index;
		if (!group[key]) {
			group[key] = [subject];
			keys.push(key);
		} else {
			group[key].push(subject);
		}
	});

	return [group, keys];
};
```
Simply put this code in separate file and load that file via `import "./bugfix/bulk-insertion-bugfix";` inside your entry point.

Worked in all use cases I had in mysql and mariandb (`ManyToOne`/`ManyToMany`/`OneToOne`) , I'm not sure what side effect reversing order will have in other databases.

**Comment by ZBAGI** [comment]
Looks like just reverting order is not enough in some cases. Here is extended version of the bugfix which orders it based on the dependencies, so it always makes sure we add/remove it in correct order:
```ts
import { Subject } from "typeorm/persistence/Subject";
import { SubjectExecutor } from "typeorm/persistence/SubjectExecutor";

// Bugfix for bulk insertion order.
// Read more: https://github.com/typeorm/typeorm/issues/2852

function addChildsRecursively(parent: Subject, subjectPool: Subject[], sortedSubjects: Subject[] = []): Subject[] {
	sortedSubjects.push(parent);
	subjectPool.filter(s => s.parentSubject == parent).forEach(child => {
		if(sortedSubjects.includes(child)) {
			throw new Error(
				"Circular dependency detected for entites: " +
				`'${child.metadata.name}' -> ` + 
				`'${sortedSubjects.find(s => s.parentSubject == child)?.metadata.name}' -> ` +
				`'${parent.metadata.name} -> '${child.metadata.name}'`);
		}

		addChildsRecursively(child, subjectPool, sortedSubjects);
	});

	return sortedSubjects;
}

function getSubjectSortedByDependency(subjects: Subject[]): Subject[] {
	const parent = subjects.find(s => s.parentSubject == null || subjects.every(ss => ss.parentSubject != s));
	if(!parent)
		throw new Error("Could not determinate root parent. Potential circular dependency for entites: " + subjects.map(s => s.metadata.name).join(", ") + ".");

	const sorted = addChildsRecursively(parent, subjects);

	// Add entites that are detached from the dependency tree
	subjects.filter(s => !sorted.includes(s)).forEach(s => {
		if(sorted.includes(s))
			return; // This entity seems to be allready included, do not create separate tree for it
		addChildsRecursively(s, subjects, sorted);
	});

	return sorted;
}

(<any>SubjectExecutor.prototype).groupBulkSubjects = function (subjects: Subject[], type: "insert" | "delete"): [{ [key: string]: Subject[] }, string[]] {
	const group: { [key: string]: Subject[] } = {};
	const keys: string[] = [];
	const groupingAllowed = type === "delete" || (<any>this).queryRunner.connection.driver.isReturningSqlSupported();

	let sorted = subjects.length > 1 ? getSubjectSortedByDependency(subjects) : subjects;
	if(type == "delete") // We delete in opposite order
		sorted = sorted.reverse();

	sorted.forEach((subject, index) => {
		const key = groupingAllowed || subject.metadata.isJunction ? subject.metadata.name : subject.metadata.name + "_" + index;
		if (!group[key]) {
			group[key] = [subject];
			keys.push(key);
		} else {
			group[key].push(subject);
		}
	});

	return [group, keys];
};
```
I will appreciate if someone could test it in other database types to see if it has some side effect there (which I doubt it will, since I cant think of scenario in which we would like to insert it or remove it in different order then based on dependencies ).
Once that is confirmed I can prepare pull request with this bugfix.

**Comment by ZBAGI** [comment]
So looks like simply reversing order is not enough in some cases. Here is better version of that bugfix which sorts subjects based on their dependencies: 
```ts
import { Subject } from "typeorm/persistence/Subject";
import { SubjectExecutor } from "typeorm/persistence/SubjectExecutor";

// Bugfix for bulk insertion order.
// Read more: https://github.com/typeorm/typeorm/issues/2852

function addChildsRecursively(parent: Subject, subjectPool: Subject[], sortedSubjects: Subject[] = []): Subject[] {
	sortedSubjects.push(parent);
	subjectPool.filter(s => s.parentSubject == parent).forEach(child => {
		if(sortedSubjects.includes(child))
			return;

		addChildsRecursively(child, subjectPool, sortedSubjects);
	});

	return sortedSubjects;
}

function getSubjectSortedByDependency(subjects: Subject[]): Subject[] {
	const parent = subjects.find(s => s.parentSubject == null) ?? subjects.find(s => subjects.every(ss => ss.parentSubject != s));
	if(!parent)
		throw new Error("Could not determinate root parent. Potential circular dependency for entites: " + subjects.map(s => s.metadata.name).join(", ") + ".");

	const sorted = addChildsRecursively(parent, subjects);

	// Add entites that are detached from the dependency tree
	subjects.filter(s => !sorted.includes(s)).forEach(s => {
		if(sorted.includes(s))
			return; // This entity seems to be allready included, do not create separate tree for it
		addChildsRecursively(s, subjects, sorted);
	});

	return sorted;
}

(<any>SubjectExecutor.prototype).groupBulkSubjects = function (subjects: Subject[], type: "insert" | "delete"): [{ [key: string]: Subject[] }, string[]] {
	const group: { [key: string]: Subject[] } = {};
	const keys: string[] = [];
	const groupingAllowed = type === "delete" || (<any>this).queryRunner.connection.driver.isReturningSqlSupported();

	let sorted = subjects.length > 1 ? getSubjectSortedByDependency(subjects) : subjects;
	if(type == "delete") // We delete in opposite order
		sorted = sorted.reverse();

	sorted.forEach((subject, index) => {
		const key = groupingAllowed || subject.metadata.isJunction ? subject.metadata.name : subject.metadata.name + "_" + index;
		if (!group[key]) {
			group[key] = [subject];
			keys.push(key);
		} else {
			group[key].push(subject);
		}
	});

	return [group, keys];
};
```
I would appreciate if someone could test it in other database types then mysql and mariandb to make sure there will not be any site effect. Personally I don't think it would, as I can't imagine situation in which order should be different then their dependencies, but there is no substitute for testing :)
Once that confirmed I will be willing to create proper pull request with this bugfix.
